activesupport/lib/active_support/core_ext/module/attribute_accessors_per_thread.rb

```diff
@@ -52,6 +52,9 @@ def #{sym}
           end
         EOS
       end
+
+      default_val = (block_given? && default.nil?) ? yield : default
+      Thread.current["attr_" + name + "_#{sym}"] = default_val unless default_val.nil?
```

Comment:
Do you mean doing something like: 

``` ruby
namespace = "attr_" + name "#{sym}"
class_eval(<<~EOS,...)
  def self.#{sym}
   Thread.current["#{namespace}"]
  end
EOS
Thread.current[namespace] = default unless default.nil?
```

I tried this approach but the problem with calling `"#{namespace}"` in `class_eval` is that that is evaluated immediately on `class_eval`. 

What happens is that if we have a class `Klass` and it's `Subclass` doing `Klass.foo` also defines `Subclass.foo` since the classes will share the same namespace `attr_Klass_foo`. However if we keep `"attr_" + name + "_#{sym}"` this is evaluated on the method call so we'll have `attr_Klass_foo` and `attr_Subclass_foo`. 

Or do you mean a different approach?